vowpalwabbit/parser.h

```diff
@@ -23,16 +23,33 @@ license as described in the file LICENSE.
 #endif
 
 #include <memory>
+#include "queue.h"
+#include "object_pool.h"
 
 struct vw;
 struct input_options;
 
+struct example_factory
+{
+  example* operator()()
+  {
+    auto new_example = new example{};
+    memset(&new_example->l, 0, sizeof(polylabel));
+    new_example->in_use = false;
+    return new_example;
+  }
+};
+
 struct parser
 {
   v_array<substring> channels;  // helper(s) for text parsing
   v_array<substring> words;
   v_array<substring> name;
 
+  std::mutex pool_lock;
+  std::unique_ptr<VW::unbounded_object_pool<example, example_factory>> example_pool;
```

Comment:
Why do we want a pointer to the example pool rather than the example pool?

Reply:
The example pool must be constructed with the default size, so it would need to be done in the constructor. The constructor is currently hidden behind a `new_parser` method. While typing this out I am realizing it would be a good change to move this logic to a constructor. I'll add another commit for this